pkg/utils/kubernetes/utils.go

```diff
@@ -0,0 +1,112 @@
+package kubernetes
+
+import (
+	"context"
+	"errors"
+	"github.com/RHsyseng/operator-utils/pkg/resource"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"strings"
+)
+
+//OnFinalize Callback function to be executed when an object is being deleted
+type OnFinalize func() error
+
+type ObjectFinalizer map[string]OnFinalize
+
+type FinalizerManager struct {
+	objects map[types.UID]ObjectFinalizer
+	Client  client.Client
+}
+
+func NewFinalizerManager(client client.Client) FinalizerManager {
+	return FinalizerManager{
+		objects: map[types.UID]ObjectFinalizer{},
+		Client:  client,
+	}
+}
+
+//RegisterFinalizer registers a finalizer function to be executed when the object is marked to be deleted.
+func (mgr *FinalizerManager) RegisterFinalizer(owner resource.KubernetesResource, name string, onFinalize OnFinalize) error {
+	if mgr.IsFinalizing(owner) {
+		return nil
+	}
+	err := validateFinalizer(name, onFinalize)
+	if err != nil {
+		return err
+	}
+	controllerutil.AddFinalizer(owner, name)
+	if err != nil {
+		return err
+	}
+	err = mgr.Client.Update(context.TODO(), owner)
+	if err != nil {
+		return err
+	}
+	_, ok := mgr.objects[owner.GetUID()]
+	if !ok {
+		mgr.objects[owner.GetUID()] = map[string]OnFinalize{}
+	}
+	mgr.objects[owner.GetUID()][name] = onFinalize
+	return nil
+}
+
+//UnregisterFinalizer removes a finalizer and updates the owner object
+func (mgr *FinalizerManager) UnregisterFinalizer(owner resource.KubernetesResource, name string) error {
+	err := validateFinalizerName(name)
+	if err != nil {
+		return err
+	}
+	obj, err := meta.Accessor(owner)
+	controllerutil.RemoveFinalizer(obj, name)
+	err = mgr.Client.Update(context.TODO(), owner)
+	if err != nil {
+		return err
+	}
+	delete(mgr.objects[obj.GetUID()], name)
+	return nil
+}
+
+//FinalizeOnDelete triggers all the finalizers registered for the given object in case the owner is being deleted
+func (mgr *FinalizerManager) FinalizeOnDelete(owner resource.KubernetesResource) error {
```

Comment:
Can we make this implicit for the user? Ideally, they register a finalizer and it should be called automatically. What if we made a reconciler chain, and instead of the client code registering their own reconciler, they would use our wrapper at https://github.com/kiegroup/kie-cloud-operator/blob/master/pkg/controller/initializer.go#L15-L16

This wrapper could then check if any of its registered finalizers has a corresponding object with a deleted timestamp, and process it if found. It could also pass the RegisterFinalizer() function to the actual reconciler, for the registering to happen.

Reply:
If I understood correctly, I think the main reason not to do it is because the finalize callback would require some context information in most cases.

Reply:
Kubernetes allows you to express your finalizer as only a string. We want to extend that to allow some logic to be mapped to that string. IMO, there should be no other context to process it. As a user, I want to say what should happen and it is reasonable that I would expect to have a `client` for my logic, so I can read/write data. But I think it makes sense for this to happen separate from the reconcile logic.

Reply:
@ruromero any further thought on this? I think we're in an uncomfortable middle point right now, where it's neither a few helper functions, nor a framework that takes care of everything for you. I could see a helper being useful where I just call it and it tells me that something is deleted, and I do what I need. This is much more than that, but if I have to check and tell it to finalize, is it worth adopting? How about my suggestion above?